internal/server/server.go

```diff
@@ -13,161 +13,116 @@ import (
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
 	}
-
-	distributionsEncoded, err := json.Marshal(distributions)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	w.Header().Set("Content-Type", "application/json")
-	w.Write(distributionsEncoded)
+	return ctx.JSON(http.StatusOK, distributions)
 }
 
-func (s *Handlers) GetArchitectures(w http.ResponseWriter, r *http.Request) {
-	archs, err := ArchitecturesForImage(r.Context().Value("distribution").(string))
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	archsEncoded, err := json.Marshal(archs)
+func (s *Handlers) GetArchitectures(ctx echo.Context, distribution string) error {
+	archs, err := ArchitecturesForImage(distribution)
 	if err != nil {
-	http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
 	}
-	w.Header().Set("Content-Type", "application/json")
-	w.Write(archsEncoded)
+	return ctx.JSON(http.StatusOK, archs)
 }
 
-func (s *Handlers) GetComposeStatus(w http.ResponseWriter, r *http.Request) {
+func (s *Handlers) GetComposeStatus(ctx echo.Context, composeId string) error {
 	socket, ok := os.LookupEnv("OSBUILD_SERVICE")
 	if !ok {
-		socket = "http://127.0.0.1:80/"
+		socket = "http://127.0.0.1:80"
 	}
-	endpoint := "compose/" + r.Context().Value("composeId").(string)
+	endpoint := "compose/" + composeId
 
-	resp, err := http.Get(socket + endpoint)
+	resp, err := http.Get(fmt.Sprintf("%s/%s", socket, endpoint))
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
 	}
 
 	var composeStatus ComposeStatus
-	json.NewDecoder(resp.Body).Decode(&composeStatus)
-	encoded, err := json.Marshal(composeStatus)
+	err = json.NewDecoder(resp.Body).Decode(&composeStatus)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
 	}
-	w.Header().Set("Content-Type", "application/json")
-	w.Write(encoded)
+	return ctx.JSON(http.StatusOK, composeStatus)
 }
 
-func (s *Handlers) ComposeImage(w http.ResponseWriter, r *http.Request) {
+func (s *Handlers) ComposeImage(ctx echo.Context) error {
 	socket, ok := os.LookupEnv("OSBUILD_SERVICE")
 	if !ok {
 		socket = "http://127.0.0.1:80/"
 	}
 
-	decoder := json.NewDecoder(r.Body)
+	decoder := json.NewDecoder(ctx.Request().Body)
 	var composeRequest cloudapi.ComposeJSONRequestBody
 	err := decoder.Decode(&composeRequest)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("Failed decoding the compose request %s", err.Error()), http.StatusBadRequest)
-		return
+		return echo.NewHTTPError(http.StatusBadRequest, "Failed decoding the compose request %s", err.Error())
```

Comment:
`NewHTTPError()` does not accept format strings. (We are all making this mistake independently.)

Please use `fmt.Sprintf()`. Also, when using the `%v` directive, you don't need to call the `Error()` method:
```suggestion
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Failed decoding the compose request %v", err))
```

Reply:
I've been using "" + err.Error(); "%v", error, "%s", err.Error(); fmt() or not intermittently, good to finally know how to do it :D